frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx

```diff
@@ -85,34 +85,33 @@ class Visualization extends PureComponent {
     computedSettings: {},
   };
 
-  UNSAFE_componentWillMount() {
-    this.transform(this.props);
+  constructor(props) {
+    super(props);
+    this.state = this.transformPropsToState(props);
   }
 
-  UNSAFE_componentWillReceiveProps(newProps) {
+  componentDidUpdate(prevProps, prevState) {
     if (
-      !isSameSeries(newProps.rawSeries, this.props.rawSeries) ||
-      !equals(newProps.settings, this.props.settings) ||
-      !equals(newProps.timelineEvents, this.props.timelineEvents) ||
+      !isSameSeries(prevProps.rawSeries, this.props.rawSeries) ||
+      !equals(prevProps.settings, this.props.settings) ||
+      !equals(prevProps.timelineEvents, this.props.timelineEvents) ||
       !equals(
-        newProps.selectedTimelineEventIds,
+        prevProps.selectedTimelineEventIds,
         this.props.selectedTimelineEventIds,
       )
     ) {
-      this.transform(newProps);
+      this.setState(this.transformPropsToState(this.props));
```

Comment:
transformPropsToState is now pure, therefore we must setState explicitly where needed